File: token.c
#include <stdlib.h>
#include <stdio.h>
#include "token.h"

const char* typesAsStrings[] = {"BEGINNING", "END", "COMMENT", "IDKYET", "NUMBER", "IDENTIFIER", "OPERATOR", "LPAREN", "RPAREN",
"LBRACK", "RBRACK", "COMMA", "PERIOD", "SEMICOLON", "ADD", "SUBTRACT", "MULTIPLY", "DIVIDE", "SINGLEQUOTE", "DOUBLEQUOTE",
"GREATERTHAN", "LESSTHAN", "EQUALS", "NOTOP", "INVALID"};

static void allocation_failed(void) {
  printf("Out of memory.\n");
  exit(0);
}

token_t* createToken(types type, char* token, int len) {
	token_t* t = malloc(sizeof(token_t));
	if(!t) {
		allocation_failed();
	}
	t->type = type;
	t->val = token;
	t->len = len;
	t->next = NULL;
	return t;
}

void deleteTokens(token_t* head) {
	while(head != NULL) {
		free(head->val);
		token_t* next = head->next;
		free(head);
		head = next;
	}
}

File: ol/lexer/token.go
package lexer

type Token struct {
	Value  string
	Type   int
	Length int
}

//declare the different token types
const (
	NUMBER = iota
	IDENTIFIER = iota
	STRING = iota

	LPAREN = iota
	RPAREN = iota
	LBRACK = iota
	RBRACK = iota
	COMMA = iota
	PERIOD = iota
	SEMICOLON = iota
	PLUS = iota
	MINUS = iota
	TIMES = iota
	SLASH = iota

	GT = iota
	GTEQ = iota
	LT = iota
	LTEQ = iota
	EQ = iota
	EQEQ = iota
	NOT = iota
	NOTEQ = iota

	TRUE = iota
	FALSE = iota

	AND = iota
	OR = iota
	IF = iota
	ELSE = iota
	FUNC = iota

	FOR = iota
	WHILE = iota

	VAR = iota

	EOF = iota
	UNRECOGNIZED = iota
)

func CreateToken(lexer *Lexer, typ int) (*Token) {

	token := Token{Value: lexer.File[lexer.Start: lexer.Position], Type: typ}

	token.Length = lexer.Position - lexer.Start

	return &token
}
